auth/src/api/handlers.rs

```diff
@@ -39,6 +39,13 @@ pub(crate) async fn post_user(
     Ok(Json(user.into()))
 }
 
+pub(crate) async fn put_user_reset_key(
+    State(user_manager): State<UserManagerState>,
+    Path(account_name): Path<AccountName>,
+) -> Result<(), Error> {
+    user_manager.reset_key(account_name).await
+}
```

Comment:
Should we rather try to get the account to reset from either the session or the api-key bearer token? This way a user can only reset their own key, and you won't need to first get your account name (which is the github-id).

Reply:
I still have to figure out the whole picture including gateway, authentication and authorization; but I was assuming that the route leading to this function would only be accessible for the users themselves (and maybe admins).

Wouldn't using the api key to reset the api key be a huge security issue? What if the api key was leaked and now someone else uses it to reset it?

Reply:
There is a question if we should only allow this through the console, but the console isn't quite ready yet. The reason I think we could allow doing it with the api-key auth is that if you steal a key and reset it, the only way to get the new key will be to login with github on the website, and you'll also not be able to make new requests with the old key if it is invalidated in the cache. But still, there could be more problems with this I am missing, so if this isn't secure we could start with only allowing reset via console (which is wip) or by admin (then using the admin guard in `auth` on the route).

So if a user is logged in now through the CLI, their api-key will be set as a Bearer token on their requests to gateway. The gateway will then proxy this request to `auth` (if cache is missed in gateway auth layer), where the api-key bearer token will be converted to a JWT with the scopes that account has. 

The JWT is returned to the `gateway` auth layer, and then the request proceeds through the gateway router with the JWT as it's Bearer token. So if your request was to deploy a service, the request will go to `deployer` which will decode this token, and the claim's scopes will be used to determine which endpoints you have permissions to call on the `deployer` router. In the `deployer` the claim will also be set as an extension on the requests to the `runtime` to load and start your service, and when your service runtime requests a database from the `provisioner` the claim scopes will again be checked to make sure you have permissions to provision a database of that type (currently anyone can provision any resource).

I realize this became a bit of a wall of text and I likely missed some details, we should probably do a proper write-up of this and add it to contributing.md/in a new "project layout" document. :sweat_smile:

Reply:
You're absolutely right about the current dashboard issues, we've been aware of them but we have been prioritizing work on the console, thinking it would be a duplicated effort since the console will arrive soon. In hindsight we should have fixed the dashboard first. :see_no_evil: 

Currently the reset key endpoint on `auth` is under `/users` so requests to it will be forward directly from `gateway` to `auth` (see `forward_to_auth` in `gateway/api/auth_layer`). My thinking is we can check for a cookie *or* an api-key in the new handler, and then get the user from that and reset the key. This way both console and CLI can use it. We'd also need to try to get the cookie sid or api-key (either can be a cache key) in the auth layer and invalidate the token if it's there in the auth layer. Since we do almost the same thing in the case of `!forward_to_auth`, we might do some refactoring of the auth layer logic to make this cleaner.

Unless you guys can see any other issues with the CLI being able to reset the token, the way I see it an intruder resetting the token will only reduce their access. With the token reset and invalidated in the cache, they won't be able to send new requests, but any jobs they started previously may complete since the token lives until expiry. The reason the JWT lives so long (15 min) is architectural constraints, so a deployer might take very long to compile a project, and then the JWT needs to live long enough after to provision resources. We should be able to improve this with the new architecture where the builder is it's own service, I think.

Reply:
Thanks, @oddgrd, for the detailed explanations. I have changed the PR and followed your advice wrt `shuttle-auth`. `gateway` still needs to be done.

Reply:
Great, the changes look good! Let us know if you need help with the rest :slightly_smiling_face: